cli.py

```diff
@@ -0,0 +1,618 @@
+    losses = AverageMeter()
+    top1 = AverageMeter()
+
+    model.train()
+
+    end = time.time()
+    ema_loss, steps = 0, 0
+    for i, (input_, target) in enumerate(train_loader):
+        steps += 1
+        total_steps += 1
+
+        if torch.cuda.is_available():
+            target = target.cuda(async=True)
+            input_ = input_.cuda()
+
+        if shape is None:
+            input_var = torch.autograd.Variable(input_)
+        else:
+            input_var = torch.autograd.Variable(input_.view(shape))
+        target_var = torch.autograd.Variable(target)
+
+        if isinstance(optimizer, VProp):
+            # Calculate noisy loss
+            optimizer.add_noise_to_parameters()
+            output = model(input_var)
+            loss = criterion(output, target_var, model)
+
+            # Do an update
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.remove_noise_from_parameters()
+            optimizer.step()
+
+            # Calculate clean loss to update metrics
+            output = model(input_var)
+            loss = criterion(output, target_var, model)
+
+        else:
+            output = model(input_var)
+            loss = criterion(output, target_var, model)
+
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+        prec1 = accuracy(output.data, target, topk=(1,))[0]
+        losses.update(loss.item(), input_.size(0))
+        top1.update(prec1.item(), input_.size(0))
+
+        if clip_var:
+            for k, layer in enumerate(model.layers):
+                layer.constrain_parameters(thres_std=thres_stds[k])
+
+        if model.beta_ema > 0.:
+            model.update_ema()
+
+        batch_time.update(time.time() - end)
+        end = time.time()
+
+        if i % print_freq == 0:
+            print(f' Epoch: [{epoch}][{i}/{len(train_loader)}]\t' +
+                  f'Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t' +
+                  f'Loss {losses.val:.4f} ({losses.avg:.4f})\t' +
+                  f'Prec@1 {top1.val:.3f} ({top1.avg:.3f})')
+
+    if writer is not None:
+        writer.add_scalar('train/loss', losses.avg, epoch)
+        writer.add_scalar('train/acc', top1.avg, epoch)
+
+    return total_steps
+
+
+def validate(val_loader: object,
+             model: torch.nn.Module,
+             criterion: object,
+             epoch: int,
+             print_freq: int,
+             writer: object) -> float:
+
+    batch_time = AverageMeter()
+    losses = AverageMeter()
+    top1 = AverageMeter()
+
+    model.eval()
+    if model.beta_ema > 0:
+        old_params = model.get_params()
+        model.load_ema_params()
+
+    end = time.time()
+    for i, (input_, target) in enumerate(val_loader):
+        if torch.cuda.is_available():
+            target = target.cuda(async=True)
+            input_ = input_.cuda()
+
+        input_var = torch.autograd.Variable(input_.view(-1, 784))
```

Comment:
You probably need to pass a shape variable here as well (similarly to the train function)